File: app/test/unit/core/engines/parity_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:lifedna/core/engines/macro_calculator.dart';
import 'package:lifedna/core/engines/recovery_engine.dart';
import 'package:lifedna/shared/enums/enums.dart';

/// ENGINE PARITY.
///
/// The Recovery and Macro engines exist twice — once in Dart on the client so
/// the UI can recompute instantly and offline, once in TypeScript on the
/// server which is authoritative for what gets stored. If they ever disagree,
/// a user sees one recovery score before syncing and a different one after.
/// That is intolerable, so parity is enforced mechanically rather than by
/// convention.
///
/// These fixtures live at the repository root and are executed by BOTH suites:
///   app/test/unit/core/engines/parity_test.dart   (this file)
///   functions/test/engines/parity.test.ts
///
/// A divergence fails both builds. If you change a formula, change it in both
/// implementations and regenerate the fixtures.
void main() {
  final fixtureRoot = Directory('../test/fixtures/engines');

  group('MacroCalculator parity', () {
    final dir = Directory('${fixtureRoot.path}/macro');

    test('fixture directory is present', () {
      expect(
        dir.existsSync(),
        isTrue,
        reason: 'Expected shared fixtures at ${dir.absolute.path}',
      );
      expect(dir.listSync().whereType<File>(), isNotEmpty);
    });

    for (final file in _fixtures(dir)) {
      final fixture =
          json.decode(file.readAsStringSync()) as Map<String, dynamic>;
      final id = fixture['id'] as String;

      test('macro/$id', () {
        final input = fixture['input'] as Map<String, dynamic>;
        final expected = fixture['expected'] as Map<String, dynamic>;

        final result = MacroCalculator.compute(
          MacroInput(
            weightKg: (input['weightKg'] as num).toDouble(),
            heightCm: (input['heightCm'] as num).toDouble(),
            age: input['age'] as int,
            sex: Sex.fromWire(input['sex'] as String),
            activityLevel:
                ActivityLevel.fromWire(input['activityLevel'] as String),
            goalMode: GoalMode.fromWire(input['goalMode'] as String),
            trainingDaysPerWeek: (input['trainingDaysPerWeek'] as int?) ?? 4,
            leanMassKg: (input['leanMassKg'] as num?)?.toDouble(),
            weeklyRateTargetPct:
                (input['weeklyRateTargetPct'] as num?)?.toDouble() ?? 0.75,
          ),
        );

        expect(result.bmr, expected['bmr'], reason: 'bmr');
        expect(result.tdee, expected['tdee'], reason: 'tdee');
        expect(result.proteinFloorG, expected['proteinFloorG'],
            reason: 'proteinFloorG');
        expect(result.waterMl, expected['waterMl'], reason: 'waterMl');
        expect(result.clamped, expected['clamped'], reason: 'clamped');
        expect(
          result.warnings,
          (expected['warnings'] as List).cast<String>(),
          reason: 'warnings',
        );
        expect(
          result.projectedWeeklyChangeKg,
          closeTo((expected['projectedWeeklyChangeKg'] as num).toDouble(), 0.01),
          reason: 'projectedWeeklyChangeKg',
        );

        _expectTargets(
          result.trainingDay,
          expected['trainingDay'] as Map<String, dynamic>,
          'trainingDay',
        );
        _expectTargets(
          result.restDay,
          expected['restDay'] as Map<String, dynamic>,
          'restDay',
        );
      });
    }
  });

  group('RecoveryEngine parity', () {
    final dir = Directory('${fixtureRoot.path}/recovery');

    test('fixture directory is present', () {
      expect(dir.existsSync(), isTrue);
      expect(dir.listSync().whereType<File>(), isNotEmpty);
    });

    for (final file in _fixtures(dir)) {
      final fixture =
          json.decode(file.readAsStringSync()) as Map<String, dynamic>;
      final id = fixture['id'] as String;

      test('recovery/$id', () {
        final input = fixture['input'] as Map<String, dynamic>;
        final expected = fixture['expected'] as Map<String, dynamic>;

        final result = RecoveryEngine.compute(
          sleep: _sleep(input['sleep'] as Map<String, dynamic>?),
          training: _training(input['training'] as Map<String, dynamic>?),
          activity: _activity(input['activity'] as Map<String, dynamic>?),
          physiology: _physiology(input['physiology'] as Map<String, dynamic>?),
          plannedSession:
              _planned(input['plannedSession'] as Map<String, dynamic>?),
        );

        expect(result.recoveryScore, expected['recoveryScore'],
            reason: 'recoveryScore');
        expect(result.band.wire, expected['band'], reason: 'band');
        expect(result.readinessScore, expected['readinessScore'],
            reason: 'readinessScore');
        expect(result.sleepScore, expected['sleepScore'], reason: 'sleepScore');
        expect(result.action?.wire, expected['action'], reason: 'action');
        expect(
          result.missingInputs,
          (expected['missingInputs'] as List).cast<String>(),
          reason: 'missingInputs',
        );

        final expectedComponents =
            (expected['components'] as List).cast<Map<String, dynamic>>();
        expect(result.components.length, expectedComponents.length,
            reason: 'component count');
        for (var i = 0; i < expectedComponents.length; i++) {
          final actual = result.components[i];
          final want = expectedComponents[i];
          expect(actual.name, want['name'], reason: 'component[$i].name');
          expect(actual.score, want['score'], reason: 'component[$i].score');
          expect(actual.weight, closeTo((want['weight'] as num).toDouble(), 0.001),
              reason: 'component[$i].weight');
          expect(
            actual.contribution,
            closeTo((want['contribution'] as num).toDouble(), 0.1),
            reason: 'component[$i].contribution',
          );
          expect(actual.available, want['available'],
              reason: 'component[$i].available');
        }

        if (expected['detail'] != null) {
          expect(result.detail, expected['detail'], reason: 'detail');
        }
      });
    }
  });
}

List<File> _fixtures(Directory dir) {
  if (!dir.existsSync()) return const [];
  return dir
      .listSync()
      .whereType<File>()
      .where((f) => f.path.endsWith('.json'))
      .toList()
    ..sort((a, b) => a.path.compareTo(b.path));
}

void _expectTargets(
  dynamic actual,
  Map<String, dynamic> want,
  String label,
) {
  expect(actual.kcal, want['kcal'], reason: '$label.kcal');
  expect(actual.proteinG, want['proteinG'], reason: '$label.proteinG');
  expect(actual.carbsG, want['carbsG'], reason: '$label.carbsG');
  expect(actual.fatG, want['fatG'], reason: '$label.fatG');
  expect(actual.proteinFloorG, want['proteinFloorG'],
      reason: '$label.proteinFloorG');
}

SleepInput? _sleep(Map<String, dynamic>? j) => j == null
    ? null
    : SleepInput(
        totalMinutes: j['totalMinutes'] as int,
        timeInBedMinutes: j['timeInBedMinutes'] as int,
        goalMinutes: (j['goalMinutes'] as int?) ?? 480,
        deepMinutes: j['deepMinutes'] as int?,
        remMinutes: j['remMinutes'] as int?,
        bedtimeStdDevMinutes:
            (j['bedtimeStdDevMinutes'] as num?)?.toDouble(),
        wakeStdDevMinutes: (j['wakeStdDevMinutes'] as num?)?.toDouble(),
        nightsOfHistory: (j['nightsOfHistory'] as int?) ?? 0,
      );

TrainingInput? _training(Map<String, dynamic>? j) => j == null
    ? null
    : TrainingInput(
        acwr: (j['acwr'] as num?)?.toDouble(),
        yesterdayLoad: (j['yesterdayLoad'] as num).toDouble(),
        meanDailyLoad28d: (j['meanDailyLoad28d'] as num).toDouble(),
        yesterdaySessionRpe: j['yesterdaySessionRpe'] as int?,
        daysSinceLastSession: (j['daysSinceLastSession'] as int?) ?? 0,
      );

ActivityInput? _activity(Map<String, dynamic>? j) => j == null
    ? null
    : ActivityInput(
        steps: j['steps'] as int,
        stepGoal: j['stepGoal'] as int,
        activeMinutes: j['activeMinutes'] as int,
        trainedYesterday: (j['trainedYesterday'] as bool?) ?? false,
      );

PhysiologyInput _physiology(Map<String, dynamic>? j) => j == null
    ? const PhysiologyInput()
    : PhysiologyInput(
        restingHrBpm: j['restingHrBpm'] as int?,
        baselineRestingHrBpm: j['baselineRestingHrBpm'] as int?,
        hrvMs: (j['hrvMs'] as num?)?.toDouble(),
        baselineHrvMs: (j['baselineHrvMs'] as num?)?.toDouble(),
      );

PlannedSession? _planned(Map<String, dynamic>? j) => j == null
    ? null
    : PlannedSession(
        rpe: j['rpe'] as int,
        durationMinutes: j['durationMinutes'] as int,
      );

File: functions/test/engines/parity.test.ts
/**
 * ENGINE PARITY — the TypeScript half.
 *
 * Executes the same fixtures as `app/test/unit/core/engines/parity_test.dart`.
 * See that file for why this duplication exists and why it is enforced
 * mechanically rather than by convention.
 */
import { readdirSync, readFileSync } from 'node:fs';
import { join } from 'node:path';

import { describe, expect, it } from 'vitest';

import {
  computeMacros,
  type MacroInput,
} from '../../src/engines/nutrition/macroCalculator';
import {
  computeRecovery,
  type ActivityInput,
  type PhysiologyInput,
  type PlannedSession,
  type SleepInput,
  type TrainingInput,
} from '../../src/engines/recovery/recoveryEngine';

const FIXTURE_ROOT = join(__dirname, '..', '..', '..', 'test', 'fixtures', 'engines');

interface Fixture<I, E> {
  id: string;
  engineVersion: string;
  input: I;
  expected: E;
}

function loadFixtures<I, E>(engine: string): Fixture<I, E>[] {
  const dir = join(FIXTURE_ROOT, engine);
  return readdirSync(dir)
    .filter((f) => f.endsWith('.json'))
    .sort()
    .map((f) => JSON.parse(readFileSync(join(dir, f), 'utf8')) as Fixture<I, E>);
}

describe('MacroCalculator parity', () => {
  const fixtures = loadFixtures<MacroInput, ReturnType<typeof computeMacros>>('macro');

  it('fixtures are present', () => {
    expect(fixtures.length).toBeGreaterThan(0);
  });

  for (const fixture of fixtures) {
    it(`macro/${fixture.id}`, () => {
      const actual = computeMacros(fixture.input);
      const want = fixture.expected;

      expect(actual.bmr).toBe(want.bmr);
      expect(actual.tdee).toBe(want.tdee);
      expect(actual.weeklyAverageTdee).toBe(want.weeklyAverageTdee);
      expect(actual.proteinFloorG).toBe(want.proteinFloorG);
      expect(actual.waterMl).toBe(want.waterMl);
      expect(actual.clamped).toBe(want.clamped);
      expect(actual.warnings).toEqual(want.warnings);
      expect(actual.projectedWeeklyChangeKg).toBeCloseTo(
        want.projectedWeeklyChangeKg,
        2,
      );
      expect(actual.trainingDay).toEqual(want.trainingDay);
      expect(actual.restDay).toEqual(want.restDay);
    });
  }
});

interface RecoveryFixtureInput {
  sleep?: SleepInput;
  training?: TrainingInput;
  activity?: ActivityInput;
  physiology?: PhysiologyInput;
  plannedSession?: PlannedSession;
}

describe('RecoveryEngine parity', () => {
  const fixtures =
    loadFixtures<RecoveryFixtureInput, ReturnType<typeof computeRecovery>>('recovery');

  it('fixtures are present', () => {
    expect(fixtures.length).toBeGreaterThan(0);
  });

  for (const fixture of fixtures) {
    it(`recovery/${fixture.id}`, () => {
      const actual = computeRecovery({
        sleep: fixture.input.sleep ?? null,
        training: fixture.input.training ?? null,
        activity: fixture.input.activity ?? null,
        physiology: fixture.input.physiology ?? {},
        plannedSession: fixture.input.plannedSession ?? null,
      });
      const want = fixture.expected;

      expect(actual.recoveryScore).toBe(want.recoveryScore);
      expect(actual.band).toBe(want.band);
      expect(actual.readinessScore).toBe(want.readinessScore);
      expect(actual.sleepScore).toBe(want.sleepScore);
      expect(actual.action).toBe(want.action);
      expect(actual.missingInputs).toEqual(want.missingInputs);
      expect(actual.components).toHaveLength(want.components.length);

      actual.components.forEach((component, i) => {
        const expectedComponent = want.components[i];
        expect(component.name).toBe(expectedComponent.name);
        expect(component.score).toBe(expectedComponent.score);
        expect(component.weight).toBeCloseTo(expectedComponent.weight, 3);
        expect(component.contribution).toBeCloseTo(
          expectedComponent.contribution,
          1,
        );
        expect(component.available).toBe(expectedComponent.available);
      });

      if (want.detail) {
        expect(actual.detail).toBe(want.detail);
      }
    });
  }
});

describe('safety ceilings are enforced by the engine, not the UI', () => {
  it('never returns a deficit steeper than 25 % of weekly-average expenditure', () => {
    const result = computeMacros({
      weightKg: 140,
      heightCm: 180,
      age: 30,
      sex: 'male',
      activityLevel: 'sedentary',
      goalMode: 'cut',
      trainingDaysPerWeek: 0,
      weeklyRateTargetPct: 5.0,
    });
    const deficit = result.tdee - result.restDay.kcal;
    expect(deficit).toBeLessThanOrEqual(result.weeklyAverageTdee * 0.25 + 1);
    expect(result.clamped).toBe(true);
  });

  it('never returns energy below the absolute floor', () => {
    for (const sex of ['male', 'female'] as const) {
      const result = computeMacros({
        weightKg: 45,
        heightCm: 155,
        age: 60,
        sex,
        activityLevel: 'sedentary',
        goalMode: 'cut',
        trainingDaysPerWeek: 0,
        weeklyRateTargetPct: 1.0,
      });
      expect(result.restDay.kcal).toBeGreaterThanOrEqual(
        sex === 'female' ? 1200 : 1500,
      );
    }
  });

  it('never emits a recovery score from fewer than two domains', () => {
    const result = computeRecovery({
      sleep: { totalMinutes: 450, timeInBedMinutes: 480 },
    });
    expect(result.recoveryScore).toBeNull();
    expect(result.band).toBe('insufficient_data');
    // The message must say what would fix it — never a bare "insufficient data".
    expect(result.detail).toContain('sleep data');
  });

  it('produces no NaN or out-of-range score across an input sweep', () => {
    for (let minutes = 0; minutes <= 900; minutes += 60) {
      for (const acwr of [0, 0.5, 1.0, 1.5, 2.0, 3.0]) {
        for (let steps = 0; steps <= 25000; steps += 5000) {
          const result = computeRecovery({
            sleep: { totalMinutes: minutes, timeInBedMinutes: minutes + 40 },
            training: {
              acwr,
              yesterdayLoad: 400,
              meanDailyLoad28d: 350,
            },
            activity: { steps, stepGoal: 10000, activeMinutes: 30 },
          });
          expect(Number.isFinite(result.recoveryScore)).toBe(true);
          expect(result.recoveryScore).toBeGreaterThanOrEqual(0);
          expect(result.recoveryScore).toBeLessThanOrEqual(100);
        }
      }
    }
  });
});
